### PR TITLE
Add ability to specify handler folder for templates

### PR DESCRIPTION
**Pull request by alexellis**
Signed-off-by: Alex Ellis (OpenFaaS Ltd) <alexellis2@gmail.com>

## Description

Add ability to specify handler folder for a template as defined in 
issue #754

## Motivation and Context

Fixes: #754

* The destination folder for the user's function / handler was
hard-coded to ./function/ or ./ for the Dockerfile template, this
patch allows users to override the value in template.yml
* Set handler_folder to a custom value such as "src"
* Updated new command for new handler path - the new command
needs to decide whether to find its templated handler code in the 
function folder or a custom folder.

## How Has This Been Tested?

Tested by updating the node template to use a handler_folder of
src, the build failed, then the Dockerfile was updated to match
the src folder instead of function, this resolved the error
and once deployed, it worked as expected.

Tested "faas-cli new" with the old cli and a template modified to use "src".
The result was an empty directory, I then updated faas-cli with
this patch and got the files populated as expected: i.e.
handler.js and package.json

To try this out:

1) Download the templates
2) Edit a template where you see "function" and replace it with something else like "lib" or "src"
3) Edit template.yml for that template and add the `handler_folder` entry
4) Build a new CLI, run `faas-cli build --shrinkwrap` and look inside build/name/
5) See that your function code was put into lib or src
6) Run a build, test the function

### CODE REVIEW

**Comment by alexellis** [comment]
I know folks are busy with their own stuff, so I wanted to ping a few reviewers to see if someone could give this a visual check before merging. Also, if you have any comments or concerns please feel free to raise them *now*.

**Comment by alexellis** [comment]
Thank you Richard

File: builder/build.go [hunk 1]
```diff
@@ -152,34 +161,41 @@ type dockerBuild struct {
...
 	}
 
 	functionPath := tempPath
+
 	if useFunction {
-		functionPath += "/function"
+		if handlerFolder == "" {
+			functionPath = path.Join(functionPath, defaultHandlerFolder)
+		} else {
+			functionPath = path.Join(functionPath, handlerFolder)
```
Comment (martindekov) [inline builder/build.go]: not sure do we want to check for prepended `/` for `handlerFolder`?

File: builder/build.go [hunk 2]
```diff
@@ -152,34 +161,41 @@ type dockerBuild struct {
...
+		fmt.Printf("Error clearing temporary build folder: %s\n", tempPath)
 		return tempPath, clearErr
 	}
 
 	functionPath := tempPath
+
 	if useFunction {
-		functionPath += "/function"
+		if handlerFolder == "" {
+			functionPath = path.Join(functionPath, defaultHandlerFolder)
```
Comment (martindekov) [inline builder/build.go]: Not sure how this even worked before given that:
```
functionPath := tempPath    (./build/somename/)
                                             ??
functionPath += "/function" (./build/somename//function)
```
(talking about the old revision)

**Comment by martindekov** [comment]
Also strange that I see that build failed, even though I see the latest commit's build succeeded

File: builder/build.go (hunk 2 above)
Comment (alexellis) [reply builder/build.go]: Yes, I've found this kind of code in a few places and I'm fixing it up to use `path.Join`, it was probably written by younger me.

**Comment by alexellis** [comment]
Fixed the build error.

File: builder/build.go (hunk 2 above)
Comment (LucasRoesler) [reply builder/build.go]: Does it make sense to replace all of the string formatting paths with path.Join?   There are a couple more like this one https://github.com/openfaas/faas-cli/pull/755/files#diff-85b0a27de88442108d89cc278530678eR168 and this one https://github.com/openfaas/faas-cli/pull/755/files#diff-85b0a27de88442108d89cc278530678eR345

Comment (alexellis) [reply builder/build.go]: I left those as they were a bit more involved with the sprint, I think path.Join takes a slice?

File: builder/build.go (hunk 1 above)
Comment (alexellis) [reply builder/build.go]: It should add it through path.Join

**Pull request merged**